src/Servers.jl

```
@@ -130,278 +109,196 @@ function check_rate_limit(tcp;
-        @show HTTP.payload(request)
+        @show HTTP.payload(request, String)
         return HTTP.Response(404)
     end
 ```
 """
-listen(f, host::IPAddr=Sockets.localhost, port::Integer=8081; kw...) = listen(f, Sockets.InetAddr(host, port); kw...)
-listen(f, host::AbstractString, port::Integer; kw...) = listen(f, parse(IPAddr, host), port; kw...)
-listen(f, host::AbstractString; kw...) = listen(f, string(host); kw...)
+function listen end
+
+const nolimit = typemax(Int)
 
-function listen(f::Function,
-                host::Union{Sockets.InetAddr, String},
+getinet(host::String, port::Integer) = Sockets.InetAddr(parse(IPAddr, host), port)
+getinet(host::IPAddr, port::Integer) = Sockets.InetAddr(host, port)
+
+function listen(f,
+                host::Union{IPAddr, String}=Sockets.localhost,
+                port::Integer=8081
                 ;
-                ssl::Bool=false,
-                require_ssl_verification::Bool=true,
-                sslconfig::Union{SSLConfig, Nothing}=nothing,
-                pipeline_limit::Int=ConnectionPool.default_pipeline_limit,
-                tcpisvalid::Function=(tcp; kw...)->true,
-                tcpref::Ref=Ref{Base.IOServer}(),
+                stream::Bool=false,
+                sslconfig::Union{MbedTLS.SSLConfig, Nothing}=nothing,
+                tcpisvalid::Union{Function, Nothing}=nothing,
+                tcpref::Union{Ref, Nothing}=nothing,
                 reuseaddr::Bool=false,
-                connectioncounter::Base.RefValue{Int}=Ref(0),
-                kw...)
+                connectioncounter::Ref{Int}=Ref(0),
+                ratelimit::Union{Rational{Int}, Nothing}=nothing,
+                reuse_limit::Int=nolimit, readtimeout::Int=0,
+                verbose::Bool=false)
 
-    if ssl && sslconfig === nothing
-        sslconfig = SSLConfig(require_ssl_verification)
-    end
+    handler = f isa Handler ? f : stream ? StreamHandlerFunction(f) : RequestHandlerFunction(f)
```

Should there be a precondition `@require !(f isa Handler) && stream == true)`.
Otherwise the `stream` argument is silently ignored.

---

Someone could pass their own `StreamHandler` that operates directly on a `Stream` (via dispatch). The `stream` keyword argument is really only for when you pass a function.